### PR TITLE
API-diff between .NET 6.0 and .NET 7.0

### PR DESCRIPTION
**Pull request by carlossanlop**
- ASP.NET - @dotnet/aspnet-api-review 
- WinForms - @merriemcgaw @RussKie
- WPF - @singhashish-wpf and @pchaurasia14
- Libraries - @jeffhandley @karelz @ericstj @stephentoub

- Microsoft.VisualBasic: @jaredpar @cston @333fred
- Microsoft.Win32: @dotnet/area-microsoft-win32
- System.Buffers: @dotnet/area-system-buffers 
- System.CodeDom: @dotnet/area-system-codedom 
- System.Collections: @dotnet/area-system-collections 
- System.ComponentModel: @dotnet/area-system-componentmodel 
- System.ComponentModel.Composition: @dotnet/area-system-componentmodel-composition 
- System.Data: @ajcvickers @DavoudEshtehari @David-Engel @JRahnama 
- System.Diagnostics.Activity: @dotnet/area-system-diagnostics-activity 
- System.Diagnostics.CodeAnalysis: @AaronRobinsonMSFT 
- System.Diagnostics.EventLog: @dotnet/area-system-diagnostics-eventlog 
- System.Diagnostics.Metrics: @tommcdon @noahfalk 
- System.Diagnostics.PerformanceCounter: @dotnet/area-system-diagnostics-performancecounter 
- System.Diagnostics.Process: @dotnet/area-system-diagnostics-process 
- System.Formats.Tar: @dotnet/area-system-io
- System.IO: @dotnet/area-system-io 
- System.IO.Compression: @dotnet/area-system-io-compression 
- System.Linq: @dotnet/area-system-linq 
- System.Net: @dotnet/ncl 
- System.Numerics: @dotnet/area-system-numerics 
- System.Reflection: @dotnet/area-system-reflection 
- System.Reflection.Emit: @dotnet/area-system-reflection-emit 
- System.Reflection.Metadata: @dotnet/area-system-reflection-metadata 
- System.Resources: @dotnet/area-system-resources 
- System.Runtime.CompilerServices: @dotnet/area-system-runtime-compilerservices 
- System.Runtime.InteropServices: @dotnet/interop-contrib @lewing @pavelsavara 
- System.Runtime.Intrinsics: @dotnet/area-system-runtime-intrinsics 
- System.Runtime.Serialization: @HongGit @StephenMolloy 
- System.Runtime.Versioning: @buyaa-n 
- System.Runtime: @dotnet/area-system-runtime 
- System.Security: @dotnet/area-system-security 
- System.Text: @dotnet/area-system-text-encoding 
- System.Text.Json: @dotnet/area-system-text-json 
- System.Text.RegularExpressions: @dotnet/area-system-text-regularexpressions 
- System.Threading: @mangod9 @kouvel 
- System.Timers: @dotnet/area-system-componentmodel 
- System.Transactions: @HongGit @roji
- System.Xml: @dotnet/area-system-xml 
- System.Web: @Rick-Anderson 
- System: @dotnet/area-system-runtime

### CODE REVIEW

**Review by RussKie — APPROVED** [review]
Windows Forms looks expected, though the diff is hard to parse.

**Review by roji — APPROVED** [review]
OK for System.Transactions

**Review by pchaurasia14 — APPROVED** [review]
No direct changes for WPF. LGTM.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.IO.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
++        public virtual void Lock(long position, long length);
+-        [UnsupportedOSPlatformAttribute("ios")]
+-        [UnsupportedOSPlatformAttribute("macos")]
+-        [UnsupportedOSPlatformAttribute("tvos")]
+-        public virtual void Unlock(long position, long length);
++        [UnsupportedOSPlatformAttribute("freebsd")]
++        [UnsupportedOSPlatformAttribute("ios")]
++        [UnsupportedOSPlatformAttribute("macos")]
++        [UnsupportedOSPlatformAttribute("tvos")]
++        public virtual void Unlock(long position, long length);
```
Comment (adamsitnik) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.IO.md:57]: Only 1 attribute was added, this diff suggests that 3 were removed and 4 were added.

```suggestion
+        [UnsupportedOSPlatformAttribute("freebsd")]
        public virtual void Lock(long position, long length);
+        [UnsupportedOSPlatformAttribute("freebsd")]
        public virtual void Unlock(long position, long length);
```

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.IO.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
++        public static bool Exists([NotNullWhenAttribute(true)] string? path);
+     }
+     public static class RandomAccess {
++        public static void SetLength(SafeFileHandle handle, long length);
+     }
+     public abstract class Stream : MarshalByRefObject, IAsyncDisposable, IDisposable {
+-        [ObsoleteAttribute("Stream.CreateWaitHandle has been deprecated. Use the ManualResetEvent(false) constructor instead.")]
+-        protected virtual WaitHandle CreateWaitHandle();
++        [ObsoleteAttribute("CreateWaitHandle has been deprecated. Use the ManualResetEvent(false) constructor instead.")]
++        protected virtual WaitHandle CreateWaitHandle();
```
Comment (adamsitnik) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.IO.md:82]: Only the text for `Obsolete` has changed, I don't believe it's worth including in API diff.

```suggestion
```

**Comment by tarekgh** [comment]
System.Diagnostics.Activity, System.Diagnostics.Metrics, and System.ComponentModel LGTM. 

I assume it is intentional to list the nullability differences in the existing APIs too. Right?

File: release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Diagnostics.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Microsoft.AspNetCore.Diagnostics
+
+``` diff
+ namespace Microsoft.AspNetCore.Diagnostics {
+     public interface IStatusCodeReExecuteFeature {
++        Endpoint Endpoint { get; }
+-        string? OriginalQueryString { get; set; }
++        string OriginalQueryString { get; set; }
```
Comment (BrennanConroy) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Diagnostics.md:8]: Wrong nullability

**Review by BrennanConroy — COMMENTED** [review]
RateLimiting looks good. Nullability in AspNetCore is off.

**Review by cston — APPROVED** [review]
Microsoft.VisualBasic LGTM.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# System.Runtime.Serialization
+
+``` diff
+ namespace System.Runtime.Serialization {
+     public class ExportOptions {
++        public ISerializationSurrogateProvider DataContractSurrogate { get; set; }
```
Comment (StephenMolloy) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.md:6]:
```suggestion
+        public ISerializationSurrogateProvider? DataContractSurrogate { get; set; }
```

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.DataContracts.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
++        public virtual ReadOnlyCollection<DataMember> DataMembers { get; }
++        public virtual bool IsBuiltInDataContract { get; }
++        public virtual bool IsISerializable { get; }
++        public virtual bool IsReference { get; }
++        public virtual bool IsValueType { get; }
++        public virtual Dictionary<XmlQualifiedName, DataContract>? KnownDataContracts { [RequiresUnreferencedCodeAttribute("Data Contract Serialization and Deserialization might require types that cannot be statically analyzed. Make sure all of the required types are preserved.")] get; }
++        public virtual Type OriginalUnderlyingType { get; }
++        public virtual XmlDictionaryString? TopLevelElementName { get; }
++        public virtual XmlDictionaryString? TopLevelElementNamespace { get; }
++        public virtual Type UnderlyingType { get; }
```
Comment (StephenMolloy) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.DataContracts.md:17]:
```suggestion
+        [DynamicallyAccessedMembers(ClassDataContract.DataContractPreserveMemberTypes)]
+        public virtual Type UnderlyingType { get; }
```

I don't know if the `DynamicallyAccessedMember` attribute is importatnt for documenting here. It appears the `RequiresUnreferencedCode` attribute is documented. The parameter for the attribute here is an internal constant.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.DataContracts.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
++        public string Name { get; }
++        public long Order { get; }
++    }
++    public sealed class XmlDataContract : DataContract {
++        public bool HasRoot { get; }
++        public bool IsAnonymous { get; }
++        public bool IsTopLevelElementNullable { get; }
++        public bool IsTypeDefinedOnImport { get; set; }
++        public new bool IsValueType { get; set; }
++        public XmlSchemaType XsdType { get; }
```
Comment (StephenMolloy) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.DataContracts.md:60]:
```suggestion
+        public XmlSchemaType? XsdType { get; }
```

**Review by StephenMolloy — APPROVED** [review]
System.Runtime.Serialization looks mostly good with just a couple suggestions.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.DataContracts.md (hunk 5 above)
Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.Serialization.DataContracts.md:17]: I'll add `DynamicallyAccessedMembers` to the list of attributes to ignore.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
++    [EventHandlerAttribute("ontouchcancel", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchend", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchenter", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchleave", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchmove", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchstart", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("onvolumechange", typeof(EventArgs), true, true)]
++    [EventHandlerAttribute("onwaiting", typeof(EventArgs), true, true)]
++    [EventHandlerAttribute("onwheel", typeof(WheelEventArgs), true, true)]
++    public static class EventHandlers
```
[the remark this answers is not in the document]
Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md:192]: Can you please elaborate?

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.IO.md (hunk 1 above)
Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.IO.md:57]: Yeah, the api-diff tool is dump in properly diffing attributes differences. It shows everything or nothing. Thanks for the suggested correction.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md:192]: Where are the `{` for EventHandlers?

[the remark this answers is not in the document]
Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md:192]: Oh I see. Not sure what happened here, it should've been added. I'll suggest it:
```suggestion
+    public static class EventHandlers {
```

File: release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
++    [EventHandlerAttribute("ontouchcancel", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchend", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchenter", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchleave", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchmove", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("ontouchstart", typeof(TouchEventArgs), true, true)]
++    [EventHandlerAttribute("onvolumechange", typeof(EventArgs), true, true)]
++    [EventHandlerAttribute("onwaiting", typeof(EventArgs), true, true)]
++    [EventHandlerAttribute("onwheel", typeof(WheelEventArgs), true, true)]
++    public static class EventHandlers {
```
Comment (BrennanConroy) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md:192]:
```suggestion
+    public static class EventHandlers {
     }
```

Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md:192]: Already added it myself. I was faster 😄 .

Comment (BrennanConroy) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md:192]: ? This is fixing your suggestiong

Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.AspNetCore.App/7.0.0_Microsoft.AspNetCore.Components.Web.md:192]: Ah, I see. I missed the closing one.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.InteropServices.JavaScript.md [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
++    [AttributeUsageAttribute(AttributeTargets.Method, Inherited=false, AllowMultiple=false)]
++    [SupportedOSPlatformAttribute("browser")]
++    public sealed class JSExportAttribute : Attribute {
++        public JSExportAttribute();
++    }
++    [CLSCompliantAttribute(false)]
++    [EditorBrowsableAttribute(EditorBrowsableState.Never)]
++    [SupportedOSPlatformAttribute("browser")]
++    public sealed class JSFunctionBinding {
++        public JSFunctionBinding();
```
Comment (pavelsavara) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.InteropServices.JavaScript.md:18]: This constructor need to be internal. Otherwise `System.Runtime.InteropServices.JavaScript` LGTM

File: release-notes/7.0/7.0.0/api-diff/Microsoft.WindowsDesktop.App/7.0.0_System.Windows.Forms.Design.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+-        public Icon Icon { get; set; }
++        [AllowNullAttribute]
++        public Icon Icon { get; set; }
+     }
+     public interface IWindowsFormsEditorService {
+-        void DropDownControl(Control control);
++        void DropDownControl(Control? control);
+     }
+     public abstract class PropertyTab : IExtenderProvider {
+-        public virtual PropertyDescriptorCollection GetProperties(ITypeDescriptorContext? context, object component, Attribute[]? attributes);
```
Comment (dreddy-work) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.WindowsDesktop.App/7.0.0_System.Windows.Forms.Design.md:30]: I do not see this change right. @carlossanlop can you double check line 30 becoming 31?  Here is what i see in 7.0 : https://github.com/dotnet/winforms/blob/35628d369b17a61dc9b564de30a7b6ae5d5285b1/src/System.Windows.Forms/src/System/Windows/Forms/Design/PropertyTab.cs#L123

File: release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.InteropServices.JavaScript.md (hunk 9 above)
Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.NETCore.App/7.0.0_System.Runtime.InteropServices.JavaScript.md:18]:
```suggestion
+        internal JSFunctionBinding();
```

File: release-notes/7.0/7.0.0/api-diff/Microsoft.WindowsDesktop.App/7.0.0_System.Windows.Forms.Design.md (hunk 10 above)
Comment (carlossanlop) [reply release-notes/7.0/7.0.0/api-diff/Microsoft.WindowsDesktop.App/7.0.0_System.Windows.Forms.Design.md:30]: The only change I see is that the return value is nullable. I'll fix it in line 31.

File: release-notes/7.0/7.0.0/api-diff/Microsoft.WindowsDesktop.App/7.0.0_System.Windows.Forms.Design.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
++        public virtual PropertyDescriptorCollection GetProperties(ITypeDescriptorContext context, object component, Attribute[]? attributes);
```
Comment (carlossanlop) [inline release-notes/7.0/7.0.0/api-diff/Microsoft.WindowsDesktop.App/7.0.0_System.Windows.Forms.Design.md:31]:
```suggestion
+        public virtual PropertyDescriptorCollection? GetProperties(ITypeDescriptorContext? context, object component, Attribute[]? attributes);
```
